### PR TITLE
[RELAND2] Change AccumulateGrad to yield `.grad`s that match weights' memory layout

### PR DESCRIPTION
**Pull request by mcarilli**
https://github.com/pytorch/pytorch/pull/40129 fixed the error responsible for the first revert, but exposed another error in the same test.

This PR is intended as the "master copy" for merge, and it runs on full CI.
Two other PRs (restricted to run on a small subset of CI) supporting debugging DDP failures/hangs with multiple devices per process (`test_c10d.py:DistributedDataParallelTest.test_grad_layout_1devicemodule_2replicaperprocess`).
- https://github.com/pytorch/pytorch/pull/40290 tries the test with purely rowmajor contiguous params on an untouched master.  In other words https://github.com/pytorch/pytorch/pull/40290 contains none of this PR's diffs aside from the test itself.
- https://github.com/pytorch/pytorch/pull/40178, for comparison, tries the test with this PR's diffs.

Both fail the same way, indicating failure is unrelated to this PR's other diffs.

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -545,7 +557,13 @@ def _sync_params(self):
...
                                                   self.modules_params[1:]):
                     for tensor, param in zip(tensors, module_params):
-                        param.set_(tensor)
+                        # Formerly, this spot used param.set_(tensor) to steal tensor's
+                        # data without a deep copy.  Unfortunately, that wiped out the
+                        # allreduce hook attached to param's AccumulateGrad function,
+                        # likely causing https://github.com/pytorch/pytorch/issues/37079.
+                        # TODO:  If set_ becomes safe to use here, use set_.
+                        # Otherwise, find another way to steal tensor's data.
+                        param.copy_(tensor)
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:566]: Thanks for fixing!

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -372,6 +377,13 @@ def model_parameters(m):
 
             for module_copy in self._module_copies[1:]:
                 for param, copy_param in zip(self.module.parameters(), parameters(module_copy)):
+                    # Reducer requires param copies have the same strides across replicas.
+                    # Fixes up copy_param strides in case replicate didn't match param strides.
+                    if param.layout is torch.strided and param.stride() != copy_param.stride():
```
Comment (mrshenli) [inline torch/nn/parallel/distributed.py:382]: Do we expect this to occur with current `replicate.py` implementation?

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -568,8 +568,11 @@ void check_gpu_tensors(const std::vector<at::Tensor>& tensors) {
     if (t.sizes() != first.sizes()) {
       throw std::runtime_error("Tensors must have identical size");
     }
-    if (!t.is_contiguous()) {
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:571]: Does this mean all ops in `ProcessGroupNCCL` no longer require contiguous? Or does `is_non_overlapping_and_dense` implies contiguous?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 4]
```diff
@@ -537,6 +620,58 @@ void Reducer::initialize_buckets(
...
+        // Specifically, we expect torch/csrc/autograd/AccumulateGrad.h produces grads
+        // that obey there "Gradient Layout Contract":
+        //   (1) if variable.is_non_overlapping_and_dense(), the stashed grad's
+        //       strides match variable.
+        //   (2) else, stashed grad is rowmajor contiguous.
+        // and create views to match.
+        //
+        // If AccumulateGrad breaks the contract, and produces a grad with an
+        // unexpected layout, performance will degrade due to poor memory access
+        // patterns when copy_ing grad data in and out of its bucket view.
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp:643]: curious, why copy_ing instead of copying?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 5]
```diff
@@ -260,8 +344,9 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
     // struct are empty, and there is no pre-existing accumulation tensor.
     // Directly assign the sparse tensor to the `contents` field.
     replica.contents = grad;
-    // The grad is not modified and dosesn't need to be written back.
-    return false;
+    replica.contents.div_(process_group_->getSize());
+    // The grad is modified in place and needs to be written back.
+    return true;
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp:349]: @pritamdamania87 could you please confirm this is still correct after #40130?

**Review by mrshenli — COMMENTED** [review]
Changes in `reducer.cpp` and `distributed.py` LGTM!

File: torch/nn/parallel/distributed.py (hunk 2 above)
Comment (mcarilli) [reply torch/nn/parallel/distributed.py:382]: Yes, replicate transposes to rowmajor contiguous as it broadcasts.  Allowing format-preserving intra-process broadcasts is nice to have but out of scope.  Code here intends to make DDP agnostic to implementation details of replicate, so we're free to change replicate (intra-process broadcast) without breaking DDP.  This only happens once on construction, so a few deep copies aren't terrible.

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (mcarilli) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:571]: > Does this mean all ops in ProcessGroupNCCL no longer require contiguous? 
Correct.  They require all inputs are non overlapping and dense with the same layout.  It turns out DDP didn't actually need this feature, but I wasn't sure that would end up true until after I had finished the DDP changes.

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 4 above)
Comment (mcarilli) [reply torch/csrc/distributed/c10d/reducer.cpp:643]: Do you mean clone()?  clone() would replace the param, which would also wipe out the AccumulateGrad hook.

Comment (mcarilli) [reply torch/csrc/distributed/c10d/reducer.cpp:643]: do you mean clone()?  We shouldn't make variable.grad a new tensor at a new location if we don't have to.  If users reference param.grad before backward, they're accustomed to that reference remaining valid after backward under certain circumstances.  See [the docstring i added](https://github.com/pytorch/pytorch/pull/40358/files#diff-d95cdc9aa4b6a8f77487f03c988006faR62-R79) for details.  The layout behavior the docstring describes is new, but the in-place vs out-of-place behavior for grad accumulation is identical to current master.  @albanD told me that my PR should avoid changing existing in-place vs out-of-place accumulation behavior to improve BC.

**Comment by mrshenli** [comment]
@albanD @ngimel Shall we get this in before branch cut?

**Comment by ngimel** [comment]
Yes, this provides important perf improvements for channels-last.

**Review by albanD — APPROVED** [review]

**Comment by ilia-cher** [comment]
fyi pytorch_linux_backward_compatibility_check_test seems to be failing (not sure if falkiness or breakage yet)
https://app.circleci.com/pipelines/github/pytorch/pytorch/184314/workflows/453434e0-4ea6-4511-8974-de67f0436745/jobs/5968549/steps

**Comment by ngimel** [comment]
It fixed itself on the next commit, and it can't be related to this PR. It could be related to merges/reverts on mkl 3d maxpooling.

**Comment by zou3519** [comment]
@mcarilli @albanD -- this PR seems to change the memory format of grads to match the weights, which seems bc-breaking due to change in semantics (but I didn't read through the PR in detail). Is that a correct assessment?

**Comment by albanD** [comment]
Hey, the original PR: https://github.com/pytorch/pytorch/pull/34904 has all the BC-breaking info in the description. We forgot to add them here as well sorry.

**Comment by zou3519** [comment]
@albanD thanks for the link to the original pr!

**Pull request closed without merging**